packages/react-router/tests/useRenderedNavigationItems.test.tsx

```diff
@@ -0,0 +1,251 @@
+import renderer from "react-test-renderer";
+import { type RootNavigationItem } from "../src/navigationItemRegistry.ts";
+import { useRenderedNavigationItems, type RenderItemFunction, type RenderSectionFunction, isNavigationLink, type NavigationLinkRenderProps, type NavigationSectionRenderProps } from "../src/useRenderedNavigationItems.tsx";
+import { useCallback, type ReactElement } from "react";
+import { renderHook } from "@testing-library/react";
+
+type RenderLinkItemFunction = (item: NavigationLinkRenderProps, index: number, level: number) => ReactElement;
+
+type RenderSectionItemFunction = (item: NavigationSectionRenderProps, index: number, level: number) => ReactElement;
+
+interface TestComponentProps {
+    navigationItems: RootNavigationItem[];
+}
+
+// Not the prettiest mock but it's easier than simpler than using createMemoryRouter and
+// it provides the required level of testing when combined with snapshot tests.
+function Link(props: Record<string, unknown>) {
+    return (
+        <div {...props} />
+    );
+}
+
+function TestComponent({ navigationItems }: TestComponentProps) {
+    const renderLink: RenderLinkItemFunction = useCallback(({ label, linkProps, additionalProps }, index, level) => {
```

Comment:
I am not sure it would makes the code easier to read.

Moving them out of the component, for me, would makes me think that those functions are utility functions rather than being tied to `TestComponent` implementation.

I would agree with that instanciating them everytime is bad but it this case I doubt it really makes a difference.